R/variable-as-methods.R

```diff
@@ -0,0 +1,143 @@
+#' is.Text(ds$v1)
+#' # [1] TRUE
+#' 
+#' # that has strings of numbers
+#' as.vector(ds$v1)
+#' # [1] "32"   "8"    "4096" "1024"
+#' 
+#' # convert this to a numeric variable with the alias `v1_numeric`
+#' ds$v1_numeric <- as.Numeric(ds$v1)
+#' 
+#' # the values are the same, but are now numerics and the type is Numeric
+#' as.vector(ds$v1_numeric)
+#' # [1]   32    8 4096 1024
+#' is.Numeric(ds$v1_numeric)
+#' # [1] TRUE
+#' 
+#' # this new variable is derived, so if new data is appended or streamed, the
+#' # new rows of data will be updated.
+#' is.derived(ds$v1_numeric)
+#' # [1] TRUE
+#' }
+#' @name variable-as-methods
+#' @aliases as.Categorical as.Datetime as.Numeric as.Text
+NULL
+
+
+#' @rdname variable-as-methods
+#' @export
+setMethod("as.Numeric", "CrunchVariable", function (x) {
+    haltIfArray(x, callingFunc = "as.Numeric()")
+    
+    # datetimes are special
+    if (is.Datetime(x)) {
+        return(zfuncExpr("datetime_to_numeric", x))
+    }
+    
+    return(zfuncExpr("cast", x, "numeric"))
+})
+
+#' @rdname variable-as-methods
+#' @export
+setMethod("as.Text", "CrunchVariable",
+          function (x, format = "%Y-%m-%d %H:%M:%S") {
+    haltIfArray(x, callingFunc = "as.Text()")
+    
+    # datetimes are special
+    if (is.Datetime(x)) {
+        return(zfuncExpr("format_datetime", x, list(value = format)))
+    }
+    
+    return(zfuncExpr("cast", x, "text"))
+})
+
+#' @rdname variable-as-methods
+#' @export
+setMethod("as.Categorical", "CrunchVariable",
+          function (x, format = "%Y-%m-%d %H:%M:%S") {
+    haltIfArray(x, callingFunc = "as.Categorical()")
+    
+    # datetimes are special
+    if (is.Datetime(x)) {
+        # no direct datetime to cat ZCL function, but can just cast the string
+        return(zfuncExpr("cast",
+                         zfunc("format_datetime", x, list(value = format)),
+                         "categorical"))
+    }
+    
+    return(zfuncExpr("cast", x, "categorical"))
+})
+
+#' @rdname variable-as-methods
+#' @export
+setMethod("as.Datetime", "CrunchVariable", function (x, format, resolution) {
+    haltIfArray(x, callingFunc = "as.Datetime()")
+    
+    # datetimes are special, so each source type must be determined
+    if (is.Numeric(x)) {
+        # TODO: validate resolution
+        return(zfuncExpr("numeric_to_datetime", x, list(value = resolution)))
+    } else if (is.Text(x) | is.Categorical(x)) {
+        return(zfuncExpr("parse_datetime", x, list(value = format)))
+    }
+    
+    halt("Can't derive Datetime variables from ", dQuote(type(x)), " variables.")
+})
+
+#' @rdname variable-as-methods
+#' @export
+as.double.CrunchVariable <- function (x, ...) as.Numeric(x)
+
+#' @rdname variable-as-methods
+#' @export
+as.character.CrunchVariable <- function (x, ...) as.Text(x)
+
+# halt if variable is an array variable. If callingFunc is provided, provide the function that
+# the user used to get to this point. 
+haltIfArray <- function (variable, callingFunc) {
```

Comment:
This is a great little function. I wonder if we want to start collecting these kinds of functions in one place, like a `utils.R` file. I feel like there are tons of them scattered around but since they are defined next to the code which they were first written to support they can be hard to find/remember.

Reply:
There is a misc.R which I moved this to.